models/vrp.rb

```diff
@@ -337,49 +337,96 @@ def self.remove_unecessary_units(hash)
+        when :vehicle
+          [:unavailable_work_day_indices, :unavailable_work_date]
+        when :schedule
+          [:unavailable_indices, :unavailable_date]
+        end
 
-      start_indice = hash[:configuration][:schedule][:range_date][:start].to_date.cwday - 1
-      end_indice = (hash[:configuration][:schedule][:range_date][:end].to_date - hash[:configuration][:schedule][:range_date][:start].to_date).to_i + start_indice
+      element[unavailable_indices] ||= []
+      element[unavailable_dates].to_a.each{ |unavailable_date|
+        element[unavailable_indices] |=
+          [(unavailable_date.to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_index]
+      }
+      element.delete(unavailable_dates)
 
-      # remove service unavailable_visit_day_date
-      [hash[:services].to_a + hash[:shipments].to_a].flatten.each{ |s|
-        next if s[:unavailable_visit_day_date].to_a.empty?
+      return unless element[:unavailable_date_ranges]
 
-        s[:unavailable_visit_day_indices] = [] unless s[:unavailable_visit_day_indices]
-        s[:unavailable_visit_day_indices] += s[:unavailable_visit_day_date].to_a.collect{ |unavailable_date|
-          (unavailable_date.to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_indice
-        }.compact
-        s.delete(:unavailable_visit_day_date)
+      element[:unavailable_index_ranges] ||= []
+      element[:unavailable_index_ranges] += element[:unavailable_date_ranges].collect{ |range|
+        {
+          start: (range[:start].to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_index,
+          end: (range[:end].to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_index,
+        }
       }
+      element.delete(:unavailable_date_ranges)
+    end
 
-      # remove vehicle unavailable_work_date
-      hash[:vehicles].each{ |vehicle|
-        next if vehicle[:unavailable_work_date].to_a.empty?
+    def self.collect_unvaible_day_indices(element, type, start_index)
+      return [] unless element[:unavailable_index_ranges].to_a.size.positive?
+
+      unavailable_indices =
+        case type
+        when :visit
+          :unavailable_visit_day_indices
+        when :vehicle
+          :unavailable_work_day_indices
+        when :schedule
+          :unavailable_indices
+        end
 
-        vehicle[:unavailable_work_day_indices] = [] unless vehicle[:unavailable_work_day_indices]
-        vehicle[:unavailable_work_day_indices] += vehicle[:unavailable_work_date].collect{ |unavailable_date|
-          (unavailable_date.to_date - hash[:configuration][:schedule][:range_date][:start]).to_i + start_indice
-        }.compact
-        vehicle.delete(:unavailable_work_date)
-      }
+      element[unavailable_indices] += element[:unavailable_index_ranges].collect{ |range|
+        (range[:start]..range[:end]).to_a
+      }.flatten.uniq.select{ |day| day >= start_index }
```

Comment:
selecting within a range is heavier than working with [range[:start], start_index].max
and then expanding the range